mathics/builtin/arithmetic.py

```diff
@@ -2003,14 +1999,33 @@ def to_sympy(self, expr, **kwargs):
             if evaluation:
                 # Min and max might be Mathics expressions. If so, evaluate them.
                 for i in (1, 2):
-                    expr = var_min_max[i]
+                    min_max_expr = var_min_max[i]
                     if not isinstance(expr, Symbol):
-                        expr = expr.evaluate(evaluation)
-                        value = expr.to_sympy(**kwargs)
+                        min_max_expr_eval = min_max_expr.evaluate(evaluation)
+                        value = min_max_expr_eval.to_sympy(**kwargs)
                         bounds[i] = value
 
-            if arg is not None and None not in bounds:
-                return sympy.summation(arg, bounds)
+            # FIXME: The below tests on SympyExpression, but really the
+            # test should be any Mathics function.
+            if isinstance(f_sympy, sympy.core.basic.Basic):
+                # sympy.summation() won't be able to handle Mathics functions in
+                # in its first argument, the function paramameter.
+                # For example in Sum[Identity[x], {x, 3}], sympy.summation can't
+                # does't evaluate Indentity[x].
+
+                if all(hasattr(i, "is_integer") and i.is_integer for i in bounds[1:]):
+                    # When we have integer bounds, it is better to not use Sympy but
+                    # use Mathics evaluation. We turn:
+                    # Sum[f[x], {<limits>}] into
+                    #   MathicsSum[Table[f[x], {<limits>}]]
+                    # where MathicsSum is self.get_result() our Iteration iterator.
```

Comment:
Bmathics suu

Reply:
I don't understand.

Reply:
Sorry, I started writing in my cellphone, then the window got closed, and it seems that a random text was sent... :)

Reply:
What I was actually wondering is about what kind of expressions we should send to sympy. Should we tag expressions that have a sympy form, to decide if send them or not?

Reply:
It is a bit of a research problem 
* when something  _needs_ to be sent to sympy
* when it _could_ work either way, and
* when it needs or is better done using Mathics Plus

Right now I know that if the bounds are not integers, those need to be sent to sympy's Sum. 

Note that just because we are using Mathics' Plus, that doesn't mean that Sympy is not involved. The terms that Plus is summing could very easily be Sympy symbolc expressions can materialized on demand, like Pi.

So in sum (pun iintended), Sympy's Sum is needed and most useful when the value is some sort of sophisticated transformation that needs to look at the entire sequence of terms in order to make a decision and there are no Mathics variables that can be expanded. For the most part we shouldn't use Sympy Sum when Mathics Plus will do the trick. Mathics Plus is also probably faster.

Sympy Sum can't expand Mathics variables that have balues because that would mean calling back Mathics to get the value and there currently is no mechanism as far as I know for doing that. 

It is this aspect that has been the source of the bugs we've been seeing.

As was recently done for Eigenvalues, a "Method" option can be added to specify which approach to use, However when no opiton was supplied need to do the best we can.

Reply:
My idea would be, add a method sympy_native():bool on the class Expression. Then, for constants (Pi, E), atoms, and numeric types we set this method to return True, and for elementary (sympy) functions, the method would return True if the arguments are all sympy_native(). Then, in operations like ``Sum``, and in functions like `to_sympy` you could check sympy_native and decide how to proceed.

Reply:
This whole area is a mess. What is needed to think this out better. Maybe coming up with a number of specific examples would help.

I think what you want to know about an expression is whether it has unbound variables, which is the more general property that matters at least in Sum.

If an unbound variable, what kind? (Sympy variable, Module local varable? global variable?)

A bound value can still be symbolic as far as sympy, numpy, or mathmp are concerned. Pi and E are bound but symbolic. (In mathmp these  are not thought of a symbolc but rather arbitrary precision real). But these kinds of values can be materialized to some approximation to sort of Numeric value in Mathics. In general it is better to keep the value unmaterialized as long a possible, so that it can be used symbolicly in mpmath, Sympy, Numpy or Mathics.

Reply:
The thing is, we could have in mathics our own set of rules to deal with unbounded sums. The true advantage of using sympy is that the effort to keep the database of conversion rules for sums (and products) are not duplicated. But this makes sense for expressions that have a direct equivalent as pure sympy expressions.  So, I guess the right algorithm would be
* Check if the iterator is finite. In that case, return Plus/@Table[expr,it]
* otherwise, check if expr.to_sympy() gives a non ``None`` value. In that case, sent it to sympy
* If not, just let mathics to evaluate it with its internal rules. If there is no rule, then leave it unevaluated.

Reply:
A lot of this and what you write before makes a lot of sense. 

Note though that: 

> * otherwise, check if expr.to_sympy() gives a non None value.

In general, this isn't good enough. Before the change expr.to_sympy() I believe was giving a non None value. The problem was however that SymPy sum had no way of materializing the iteration value of Sum. 

That said, since the step is tried second after the first succeeds, in practice will work, although it isn't bug free.

I've run out of time I can work on this in the short term. If you are up to it and want to continue please do!